libwild/src/elf_writer.rs

```diff
@@ -1658,12 +1671,19 @@ impl<'data> ObjectLayout<'data> {
             .relocation_statistics
             .get(section.part_id.output_section_id())
             .fetch_add(relocations.len() as u64, Relaxed);
-        for rel in relocations {
+        for (i, rel) in relocations.iter().enumerate() {
             if modifier == RelocationModifier::SkipNextRelocation {
                 modifier = RelocationModifier::Normal;
                 continue;
             }
             let offset_in_section = rel.r_offset.get(LittleEndian);
+            // The iterator is used for e.g. R_RISCV_PCREL_HI20 & R_RISCV_PCREL_LO12_I pair of relocations where the later
+            // one actually points to a label of the HI20 relocations and thus we need to find it. The relocation is typically
+            // right before the LO12_* relocation.
+            let relocations_to_search = relocations[..i]
```

Comment:
When I benchmark a build of this PR against main, I see about a 6% increase in instruction count. The effect on walltime is less (about 2.9-4.6%). I'm not sure what the cause is. One possibility is just from having more branches in some of our large match statements. If that's the case, then it's likely hard to do much about at present. The other possibility is some bit of code in a common path like this code here.

Do you see a small slow down too?

If it is this bit of code here - perhaps unlikely, but possible - then the two options that come to mind are (a) pass a closure that produces this iterator or (b) pass `relocations` and `i`.

Reply:
> Do you see a small slow down too?

No, I've just tried building `mold` both w/ and w/o debug info, and the speed might be 1% slower now, not more.

> If it is this bit of code here - perhaps unlikely, but possible - then the two options that come to mind are (a) pass a closure that produces this iterator or (b) pass `relocations` and `i`.

Yeah, I can experiment with that. It's true we currently build the iterator for every relocation, where in most cases, it's unused.

Reply:
Yesterday, I spent some time with the bisection and I can confirm the instruction count increase with this change, where I managed to identify 2 steps of the growth:

1. Addition of handling of `AbsoluteAddition`/`AbsoluteSubtraction`: 29.6G -> 30.5G insns
1. Addition of  `get_pair_subtraction_relocation_value` to `apply_debug_relocation`: 31.7G -> 32.1 insns

Which is kind of expected, given the logic was added to the hottest functions. Moreover, I noticed the switch from anyhow to the custom error time increased the instruction count by ~0.7G (noticed that because I started merging `main` branch at some point).